common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -117,76 +119,445 @@ public Collation(
+      }
+    }
+
+    private static class CollationSpecICU extends CollationSpec {
+
+      private static final String[] ICULocaleNames;
+      private static final Map<String, ULocale> ICULocaleMap = new HashMap<>();
+      private static final Map<String, String> ICULocaleMapUppercase = new HashMap<>();
+      private static final Map<String, Integer> ICULocaleToId = new HashMap<>();
+      private static final String ICUCollatorVersion = "153.120.0.0";
+
+      static {
+        ICULocaleMap.put("UNICODE", ULocale.ROOT);
+        ULocale[] locales = Collator.getAvailableULocales();
+        for (ULocale locale : locales) {
+          if (locale.getVariant().isEmpty()) {
+            String language = locale.getLanguage();
+            assert (!language.isEmpty());
+            StringBuilder builder = new StringBuilder(language);
+            String script = locale.getScript();
+            if (!script.isEmpty()) {
+              builder.append('_');
+              builder.append(script);
+            }
+            String country = locale.getISO3Country();
+            if (!country.isEmpty()) {
+              builder.append('_');
+              builder.append(country);
+            }
+            String localeName = builder.toString();
+            assert (!ICULocaleMap.containsKey(localeName));
+            ICULocaleMap.put(localeName, locale);
+          }
+        }
+        for (String localeName : ICULocaleMap.keySet()) {
+          String localeUppercase = localeName.toUpperCase();
+          assert (!ICULocaleMapUppercase.containsKey(localeUppercase));
+          ICULocaleMapUppercase.put(localeUppercase, localeName);
+        }
+        ICULocaleNames = ICULocaleMap.keySet().toArray(new String[0]);
+        Arrays.sort(ICULocaleNames);
+        assert (ICULocaleNames.length <= (1 << 16));
+        for (int i = 0; i < ICULocaleNames.length; i++) {
+          ICULocaleToId.put(ICULocaleNames[i], i);
+        }
+      }
+
+      public static final int UNICODE_COLLATION_ID =
+        new CollationSpecICU("UNICODE", CaseSensitivity.CS, AccentSensitivity.AS,
+          CaseConversion.UNSPECIFIED).getCollationId();
+      public static final int UNICODE_CI_COLLATION_ID =
+        new CollationSpecICU("UNICODE", CaseSensitivity.CI, AccentSensitivity.AS,
+          CaseConversion.UNSPECIFIED).getCollationId();
+
+      private CollationSpecICU(String locale, CaseSensitivity caseSensitivity,
+          AccentSensitivity accentSensitivity, CaseConversion caseConversion) {
+        super(locale, caseSensitivity, accentSensitivity, caseConversion);
+      }
+
+      public static int collationNameToId(
+          String originalName, String collationName) throws SparkException {
+        int lastPos = -1;
+        for (int i = 1; i <= collationName.length(); i++) {
+          String localeName = collationName.substring(0, i);
+          if (ICULocaleMapUppercase.containsKey(localeName)) {
+            lastPos = i;
+          }
```

Comment:
Actually that's not true, we can have `en` and `en_USA` locales. The idea is that if we have match for multiple locales, only the longest one will work because specifiers are designed to be different from any locale part. For example, collation `en_USA_CI` can match `en` locale but `USA` is invalid specifier, thus we go with `en_USA`.

Reply:
Resolving discussion, added clarification comment which explains how specifiers are not allowed to be equal to country code, resulting in property that only valid match is the longest one.